File: src/paketKritik.ts
/**
 * Kritik-Linse für das Paket: Godfiles, Logikmischung, Artefakte, Doku-Lügen.
 *
 * Gegenstück: `tauri-app/src-tauri/src/paketkritik.rs`. Zahlen ohne
 * Tausendertrenner, damit beide Oberflächen dieselbe Datei erzeugen.
 * Nur LOC + Regex + Pfadlisten – kein AST.
 */
import { GescannteDatei } from './scanner';
import { SchemaMeta } from '@propsa/core';
import { kopf } from './paketBasis';

const GOD_GRENZE = 300;

const DOMAENEN = [
  'src/components',
  'src/simulation',
  'src/render',
  'src/bus',
  'src/persistence',
] as const;

const ARTEFAKTE: Array<{ muster: string; test: (p: string) => boolean }> = [
  { muster: 'context.md', test: p => p === 'context.md' || p.endsWith('/context.md') },
  { muster: 'context.json', test: p => p === 'context.json' || p.endsWith('/context.json') },
  { muster: 'package-lock.json', test: p => p.endsWith('package-lock.json') },
  { muster: '*.tsbuildinfo', test: p => p.endsWith('.tsbuildinfo') },
  { muster: '*.bak', test: p => p.endsWith('.bak') },
  { muster: 'playwright-report/**', test: p => p.includes('playwright-report/') },
  { muster: 'test-results/**', test: p => p.includes('test-results/') },
  { muster: 'git-noir/.tmp/**', test: p => p.includes('git-noir/.tmp/') || p.includes('.tmp/') },
  { muster: '.tmp/**', test: p => /(^|\/)\.tmp\//.test(p) },
];

function dateiDomain(pfad: string): string | null {
  for (const d of DOMAENEN) if (pfad.startsWith(d + '/') || pfad === d) return d;
  return null;
}

function importDomain(spec: string): string | null {
  for (const d of DOMAENEN) {
    const kurz = d.split('/').pop()!;
    if (spec.includes(d) || spec.includes('/' + kurz + '/') || spec.includes(kurz + '/')) return d;
  }
  return null;
}

function prozent(teil: number, gesamt: number): string {
  const b = gesamt > 0 ? gesamt : 1;
  return `${Math.round((teil / b) * 100)} %`;
}

export function kritik(meta: SchemaMeta, dateien: GescannteDatei[]): string {
  const gesamtZeilen = dateien.reduce((s, d) => s + d.zeilen, 0);
  const vorhanden = new Set(dateien.map(d => d.relativerPfad));

  let text = kopf('Kritik', meta, [
    '**Hinweis:** Diese Linse bewertet Struktur, nicht Funktionalität.',
  ]);

  // --- 1. Godfiles
  text += '\n## Godfiles (> 300 Zeilen)\n\n';
  const godfiles = [...dateien].filter(d => d.zeilen > GOD_GRENZE).sort((a, b) => b.zeilen - a.zeilen);
  if (godfiles.length === 0) {
    text += 'Keine Datei über 300 Zeilen – kompakte Module.\n';
  } else {
    text += '| Datei | Sprache | Zeilen | Anteil |\n|---|---|---|---|\n';
    for (const d of godfiles) {
      text += `| \`${d.relativerPfad}\` | ${d.sprache} | ${d.zeilen} | ${prozent(d.zeilen, gesamtZeilen)} |\n`;
    }
    text += `\n${godfiles.length} Datei(en) über 300 Zeilen – Kandidaten für Schnitte.\n`;
  }

  // --- 2. Logikmischung
  text += '\n## Logikmischung (≥ 3 fremde Domänen)\n\n';
  const mischungen: Array<{ pfad: string; eigen: string; fremde: string[] }> = [];
  const importRe = /import\s+.*?\s+from\s+['\"]([^'\"]+)['\"]/g;
  for (const d of dateien) {
    const eigen = dateiDomain(d.relativerPfad);
    if (!eigen) continue;
    const fremde = new Set<string>();
    let m: RegExpExecArray | null;
    importRe.lastIndex = 0;
    while ((m = importRe.exec(d.inhalt)) !== null) {
      const dom = importDomain(m[1]);
      if (dom && dom !== eigen) fremde.add(dom);
    }
    if (fremde.size >= 3) mischungen.push({ pfad: d.relativerPfad, eigen, fremde: [...fremde].sort() });
  }
  if (mischungen.length === 0) {
    text += 'Keine Datei importiert aus ≥ 3 fremden Domänen.\n';
  } else {
    text += '| Datei | Eigen | Fremde Domänen |\n|---|---|---|\n';
    for (const r of mischungen.sort((a, b) => a.pfad.localeCompare(b.pfad))) {
      text += `| \`${r.pfad}\` | \`${r.eigen}\` | ${r.fremde.map(f => `\`${f}\``).join(', ')} |\n`;
    }
    text += `\nEmpfehlung: Domänengrenzen schärfen oder Fassade einziehen.\n`;
  }

  // --- 3. Repo-Sauberkeit
  text += '\n## Repo-Sauberkeit (Artefakte im Paket)\n\n';
  const artefakte = dateien.filter(d => ARTEFAKTE.some(a => a.test(d.relativerPfad)));
  if (artefakte.length === 0) {
    text += 'Keine Artefakte im Paket.\n';
  } else {
    const artefaktZeilen = artefakte.reduce((s, d) => s + d.zeilen, 0);
    text += `**${artefakte.length} Datei(en), ${artefaktZeilen} Zeilen (${prozent(artefaktZeilen, gesamtZeilen)}) sind Artefakte.**\n\n`;
    text += '| Datei | Zeilen | Muster |\n|---|---|---|\n';
    for (const d of [...artefakte].sort((a, b) => b.zeilen - a.zeilen)) {
      const muster = ARTEFAKTE.find(a => a.test(d.relativerPfad))!.muster;
      text += `| \`${d.relativerPfad}\` | ${d.zeilen} | \`${muster}\` |\n`;
    }
    text += '\nEmpfehlung: Muster in `IGNORIERTE_VERZEICHNISSE` / `STANDARD_AUSSCHLUESSE` aufnehmen oder via `-e` ausschließen.\n';
    if (artefakte.some(d => d.relativerPfad.endsWith('context.md') || d.relativerPfad.endsWith('context.json'))) {
      text += 'Hinweis: Das Paket enthält sein eigenes Ergebnis (`context.md`/`context.json`) – Selbstreferenz.\n';
    }
  }

  // --- 4. Doku-Wahrheit
  //
  // Eine Referenz zählt nur als „fehlend“, wenn sie wie eine echte Datei
  // aussieht (mit Endung) und kein Eintrag im `dateien`-Array passt – nicht
  // schon, wenn sie nicht wörtlich im Paket steht. Verzeichnis- und
  // Pseudo-Referenzen (`src/…`, Wildcards, Endpunkte ohne Endung) sind
  // keine Behauptung über eine Datei.
  text += '\n## Doku-Wahrheit (Pfad-Referenzen in docs/**/*.md)\n\n';
  const docs = dateien.filter(d => d.relativerPfad.startsWith('docs/') && d.relativerPfad.endsWith('.md'));
  const pfadRe = /\b(src\/[^\s\"'`\)\]]+)/g;
  const erwaehnt = new Set<string>();
  for (const d of docs) {
    let m: RegExpExecArray | null;
    pfadRe.lastIndex = 0;
    while ((m = pfadRe.exec(d.inhalt)) !== null) {
      erwaehnt.add(m[1].replace(/[.,;:]$/, ''));
    }
  }
  // Referenz ist erfüllt, wenn ein echter Paket-Eintrag exakt passt oder
  // als Verzeichnis den referenzierten Pfad abdeckt.
  const referenzErfuellt = (p: string): boolean =>
    [...vorhanden].some(v => v === p || v.startsWith(p + '/') || p.startsWith(v + '/'));
  const istDateireferenz = (p: string): boolean =>
    /\.[A-Za-z0-9]+$/.test(p) && !p.includes('*');
  const fehlend: string[] = [];
  for (const p of [...erwaehnt].sort()) {
    if (!istDateireferenz(p) || p.includes('*')) continue;
    if (!referenzErfuellt(p)) fehlend.push(p);
  }
  if (docs.length === 0) {
    text += 'Keine `docs/**/*.md` im Paket – nichts zu prüfen.\n';
  } else if (erwaehnt.size === 0) {
    text += `${docs.length} Doku-Datei(en) ohne \`src/…\`-Pfadreferenz – nichts zu prüfen.\n`;
  } else if (fehlend.length === 0) {
    text += `Alle ${erwaehnt.size} Datei-Referenzen existieren im Paket.\n`;
  } else {
    text += `**${fehlend.length} von ${erwaehnt.size} Referenzen fehlen im Paket:**\n\n`;
    for (const p of fehlend.slice(0, 30)) text += `- \`${p}\`\n`;
    if (fehlend.length > 30) text += `- … und ${fehlend.length - 30} weitere\n`;
  }

  return text;
}

File: tauri-app/src-tauri/src/paketkritik.rs
//! Kritik-Linse: Godfiles, Logikmischung, Artefakte, Doku-Lügen.

use crate::kritik_regeln::{artefakt_muster, datei_domain, import_domain, prozent, GOD_GRENZE};
use crate::scan::{DateiInfo, ScanErgebnis};
use std::collections::{BTreeSet, HashSet};

fn kopf(titel: &str, scan: &ScanErgebnis, zusatz: &[String]) -> String {
    let mut t = format!("# PROPSA – {}\n\n", titel);
    t.push_str(&format!("**Projekt:** {}\n", scan.titel));
    t.push_str(&format!("**Erzeugt:** {}\n", scan.zeitstempel));
    for z in zusatz {
        t.push_str(&format!("{}\n", z));
    }
    t
}

pub fn kritik(scan: &ScanErgebnis) -> String {
    let gesamt_zeilen: usize = scan.dateien.iter().map(|d| d.zeilen).sum();
    let vorhanden: HashSet<&str> =
        scan.dateien.iter().map(|d| d.relativer_pfad.as_str()).collect();

    let mut text = kopf(
        "Kritik",
        scan,
        &["**Hinweis:** Diese Linse bewertet Struktur, nicht Funktionalität.".to_string()],
    );

    // 1. Godfiles
    text.push_str("\n## Godfiles (> 300 Zeilen)\n\n");
    let mut godfiles: Vec<&DateiInfo> =
        scan.dateien.iter().filter(|d| d.zeilen > GOD_GRENZE).collect();
    godfiles.sort_by(|a, b| b.zeilen.cmp(&a.zeilen));
    if godfiles.is_empty() {
        text.push_str("Keine Datei über 300 Zeilen – kompakte Module.\n");
    } else {
        text.push_str("| Datei | Sprache | Zeilen | Anteil |\n|---|---|---|---|\n");
        for d in &godfiles {
            text.push_str(&format!(
                "| `{}` | {} | {} | {} |\n",
                d.relativer_pfad,
                d.sprache,
                d.zeilen,
                prozent(d.zeilen, gesamt_zeilen)
            ));
        }
        text.push_str(&format!(
            "\n{} Datei(en) über 300 Zeilen – Kandidaten für Schnitte.\n",
            godfiles.len()
        ));
    }

    // 2. Logikmischung
    text.push_str("\n## Logikmischung (≥ 3 fremde Domänen)\n\n");
    let mut mischungen: Vec<(String, String, Vec<String>)> = Vec::new();
    for d in &scan.dateien {
        if let Some(eigen) = datei_domain(&d.relativer_pfad) {
            let mut fremde = BTreeSet::new();
            for teil in d.inhalt.split("from ") {
                let rest = teil.trim().strip_prefix('\"').or_else(|| teil.trim().strip_prefix('\''));
                if let Some(rest) = rest {
                    let spec = rest.split(['\"', '\'']).next().unwrap_or("");
                    if let Some(dom) = import_domain(spec) {
                        if dom != eigen {
                            fremde.insert(dom.to_string());
                        }
                    }
                }
            }
            if fremde.len() >= 3 {
                mischungen.push((d.relativer_pfad.clone(), eigen.to_string(), fremde.into_iter().collect()));
            }
        }
    }
    if mischungen.is_empty() {
        text.push_str("Keine Datei importiert aus ≥ 3 fremden Domänen.\n");
    } else {
        mischungen.sort_by(|a, b| a.0.cmp(&b.0));
        text.push_str("| Datei | Eigen | Fremde Domänen |\n|---|---|---|---|\n");
        for (pfad, eigen, fremde) in &mischungen {
            let f = fremde.iter().map(|x| format!("`{}`", x)).collect::<Vec<_>>().join(", ");
            text.push_str(&format!("| `{}` | `{}` | {} |\n", pfad, eigen, f));
        }
        text.push_str("\nEmpfehlung: Domänengrenzen schärfen oder Fassade einziehen.\n");
    }

    // 3. Repo-Sauberkeit
    text.push_str("\n## Repo-Sauberkeit (Artefakte im Paket)\n\n");
    let mut artefakte: Vec<(&DateiInfo, &str)> = Vec::new();
    for d in &scan.dateien {
        if let Some(m) = artefakt_muster(&d.relativer_pfad) {
            artefakte.push((d, m));
        }
    }
    if artefakte.is_empty() {
        text.push_str("Keine Artefakte im Paket.\n");
    } else {
        let sum: usize = artefakte.iter().map(|(d, _)| d.zeilen).sum();
        text.push_str(&format!(
            "**{} Datei(en), {} Zeilen ({}) sind Artefakte.**\n\n",
            artefakte.len(),
            sum,
            prozent(sum, gesamt_zeilen)
        ));
        artefakte.sort_by(|a, b| b.0.zeilen.cmp(&a.0.zeilen));
        text.push_str("| Datei | Zeilen | Muster |\n|---|---|---|---|\n");
        for (d, m) in &artefakte {
            text.push_str(&format!("| `{}` | {} | `{}` |\n", d.relativer_pfad, d.zeilen, m));
        }
        text.push_str(
            "\nEmpfehlung: Muster in `IGNORIERTE_VERZEICHNISSE` / `STANDARD_AUSSCHLUESSE` aufnehmen oder via `-e` ausschließen.\n",
        );
        if artefakte.iter().any(|(d, _)| {
            d.relativer_pfad.ends_with("context.md") || d.relativer_pfad.ends_with("context.json")
        }) {
            text.push_str("Hinweis: Das Paket enthält sein eigenes Ergebnis (`context.md`/`context.json`) – Selbstreferenz.\n");
        }
    }

    // 4. Doku-Wahrheit
    text.push_str("\n## Doku-Wahrheit (Pfad-Referenzen in docs/**/*.md)\n\n");
    let docs: Vec<&DateiInfo> = scan
        .dateien
        .iter()
        .filter(|d| d.relativer_pfad.starts_with("docs/") && d.relativer_pfad.ends_with(".md"))
        .collect();
    let mut erwaehnt = BTreeSet::new();
    for d in &docs {
        let mut i = 0;
        let b = d.inhalt.as_bytes();
        while i + 4 < b.len() {
            if &b[i..i + 4] == b"src/" {
                let mut j = i + 4;
                while j < b.len() && !matches!(b[j], b' ' | b'\n' | b'\r' | b'"' | b'\'' | b')' | b']') {
                    j += 1;
                }
                let p = d.inhalt[i..j].trim_end_matches(&[',', '.', ';', ':'][..]).to_string();
                if !p.is_empty() {
                    erwaehnt.insert(p);
                }
                i = j;
            } else {
                i += 1;
            }
        }
    }
    if docs.is_empty() {
        text.push_str("Keine `docs/**/*.md` im Paket – nichts zu prüfen.\n");
    } else if erwaehnt.is_empty() {
        text.push_str(&format!("{} Doku-Datei(en) ohne `src/…`-Pfadreferenz – nichts zu prüfen.\n", docs.len()));
    } else {
        // Eine Referenz zählt nur als „fehlend“, wenn sie wie eine echte
        // Datei aussieht (mit Endung) und kein Eintrag im `dateien`-Array
        // passt – nicht schon, wenn sie nicht wörtlich im Paket steht.
        let referenz_erfuellt = |p: &str| {
            vorhanden.iter().any(|v| {
                *v == p
                    || v.starts_with(&format!("{}/", p))
                    || p.starts_with(&format!("{}/", v))
            })
        };
        let ist_datei_referenz = |p: &str| {
            !p.contains('*')
                && p.rsplit('.').next().map_or(false, |endung| {
                    !endung.is_empty() && endung.len() <= 6 && endung.chars().all(|c| c.is_ascii_alphanumeric())
                })
                && p.contains('/')
        };
        let mut fehlend: Vec<String> = Vec::new();
        for p in &erwaehnt {
            if !ist_datei_referenz(p) {
                continue;
            }
            if !referenz_erfuellt(p) {
                fehlend.push(p.clone());
            }
        }
        if fehlend.is_empty() {
            text.push_str(&format!("Alle {} Datei-Referenzen existieren im Paket.\n", erwaehnt.len()));
        } else {
            fehlend.sort();
            text.push_str(&format!("**{} von {} Referenzen fehlen im Paket:**\n\n", fehlend.len(), erwaehnt.len()));
            for p in fehlend.iter().take(30) {
                text.push_str(&format!("- `{}`\n", p));
            }
            if fehlend.len() > 30 {
                text.push_str(&format!("- … und {} weitere\n", fehlend.len() - 30));
            }
        }
    }

    text
}
